src/transport_ofi.h

```diff
@@ -833,28 +837,39 @@ void shmem_transport_cswap(shmem_transport_ctx_t* ctx, void *target, const void
     uint8_t *addr;
 
     shmem_transport_ofi_get_mr(target, pe, &addr, &key);
-
     shmem_internal_assert(len <= sizeof(double _Complex));
     shmem_internal_assert(SHMEM_Dtsize[datatype] == len);
 
+    struct fi_ioc resultv = { .addr = dest, .count = 1 };
+    const struct fi_ioc sourcev = { .addr = (void *) source, .count = 1 };
+    const struct fi_ioc comparev = { .addr = (void *) operand, .count = 1 };
+    const struct fi_rma_ioc rmav= { .addr = (uint64_t) addr, .count = 1, .key = key };
+    const struct fi_msg_atomic msg = {
+                                 .msg_iov       = &sourcev,
+                                 .desc          = NULL,
+                                 .iov_count     = 1,
+                                 .addr          = GET_DEST(dst),
+                                 .rma_iov       = &rmav,
+                                 .rma_iov_count = 1,
+                                 .datatype      = datatype,
+                                 .op            = FI_CSWAP,
+                                 .context       = NULL,
+                                 .data          = 0
+                               };
+
     SHMEM_TRANSPORT_OFI_CTX_LOCK(ctx);
     SHMEM_TRANSPORT_OFI_CNTR_INC(&ctx->pending_get_cntr);
 
     do {
-        ret = fi_compare_atomic(ctx->ep,
-                                source,
-                                1,
-                                NULL,
-                                operand,
-                                NULL,
-                                dest,
-                                NULL,
-                                GET_DEST(dst),
-                                (uint64_t) addr,
-                                key,
-                                datatype,
-                                FI_CSWAP,
-                                NULL);
+        ret = fi_compare_atomicmsg(ctx->ep,
+                                   &msg,
+                                   &comparev,
+                                   NULL,
+                                   1,
+                                   &resultv,
+                                   NULL,
+                                   1,
+                                   FI_INJECT);
     } while (try_again(ctx, ret, &polled));
```

Comment:
Just to understand, the `fi_*msg` routine call is the difference between the non-blocking and blocking routine?

Reply:
Yes and also the nonblocking routine uses `FI_INJECT` whereas the blocking routine does not.   The blocking routine /must/ be completed before returning to the application because we use temporary buffering on the stack.